lib/Frontend/CompileJobCacheKey.cpp

```diff
@@ -0,0 +1,83 @@
+//===--- CompileJobCacheKey.cpp - compile cache key methods ---------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// This file contains utility methods for creating compile job cache keys.
+//
+//===----------------------------------------------------------------------===//
+
+#include <swift/Frontend/CompileJobCacheKey.h>
+#include <swift/Basic/Version.h>
+#include <llvm/ADT/SmallString.h>
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/CAS/HierarchicalTreeBuilder.h"
+#include "llvm/CAS/ObjectStore.h"
+
+using namespace swift;
+
+// TODO: Rewrite this into CASNodeSchema.
+llvm::Expected<llvm::cas::ObjectRef> swift::createCompileJobBaseCacheKey(
+    llvm::cas::ObjectStore &CAS, ArrayRef<const char *> Args) {
+  SmallString<256> CommandLine;
+
+  // TODO: Improve this list.
+  static const std::vector<std::string> removeArgAndNext = {
+      "-o", "-supplementary-output-file-map", "-serialize-diagnostics-path",
+      "-num-threads", "-cas-path"};
+
+  // Don't count the `-frontend` in the first location since only frontend
+  // invocation can have a cache key.
+  if (Args.size() > 1 && StringRef(Args.front()) == "-frontend")
+    Args = Args.drop_front();
```

Comment:
Is this not mutating the same underlying storage as was passed to `swift::performFrontend`?